concent_api/core/subtask_helpers.py

```diff
@@ -280,17 +280,24 @@ def are_keys_and_addresses_unique_in_message_subtask_results_accepted(
     )
 
 
-def are_subtask_results_accepted_messages_signed_by_the_same_requestor(
+def are_subtask_results_accepted_messages_properly_signed(
     subtask_results_accepted_list: List[SubtaskResultsAccepted]
 ) -> bool:
     requestor_public_key = subtask_results_accepted_list[0].task_to_compute.requestor_public_key
```

Comment:
Only one small question- is it not good place for assert, that all `subtask_results_accepted` are for one requestor? I know they should be and you hate assertions, but if we get this key from one message I suggest to make 100% that all keys are the same. I can avoid hours of debugging

Reply:
Yeah. You're making an assumption that all the keys are the same (and the function will not give you the right answer if they are not) but you're not checking that. Probably some other part of the code already checks that but since this one depends on it, there should be an `assert` here.

Reply:
it **is** checked:
```python
if not (
        are_keys_and_addresses_unique_in_message_subtask_results_accepted(client_message.subtask_results_accepted_list) and
        are_subtask_results_accepted_messages_signed_by_the_same_requestor_or_concent(client_message.subtask_results_accepted_list)
    ):
```

We don't want to have code duplication.

Reply:
1) Avoiding duplication with `assert` makes no sense. It's as if you said that having unit tests is unnecessary duplication.

    There's a simple way to tell that it's not duplication. If you have two duplicates then you should be able to use one in place of the other. Otherwise they're not really the same. So we should be able to remove the validation and leave the `assert` while keeping the code functionally the same. That's not the case here. We can't because they do different things. Validation checks your input. `assert` checks your code.

    The purpose of this `assert` is to tell the programmer (and the Python interpreter) "this function will not give you the correct answer unless this condition is true". Adding this kind of meta-information to your function does not help you validate the input. It helps you detect that someone is using it wrong. Or that someone broke some external mechanism that this function cannot properly work without (i.e. the input validation).

2) Actually, with the new name the assertion is no longer necessary in my opinion.

    When it was called `are_subtask_results_accepted_messages_properly_signed()` I would be confused if it told me that messages are properly signed even though every one has a different requestor key inside. You could avoid that by checking it inside the function with an `if`. But **that** would be duplication. The third alternative is to just say "this case is not supported; never call this function with that input" with `assert`. Problem solved.

    But now that it's called `are_subtask_results_accepted_messages_signed_by_the_same_requestor_or_concent()`, it's fine because it does exactly what it says. It only checks if the signatures are the same but does not guarantee that each message is signed with the key inside it.

3) So `are_keys_and_addresses_unique_in_message_subtask_results_accepted()` checks if the addresses and keys in all messages are the same? That's confusing. If all elements on a list are unique, it means that they're all different and that there are no duplicates. That's how `unique` works in in Django models for example. It was not a part of this pull request, but I think we should rename it (and the local variables inside it).